### PR TITLE
Feature/paired class

### PR DESCRIPTION
**Pull request by dpressel**
Datasets like NLI and paraphrase are commonly treated as classifiers with paired input.  There are lots of methods that could be used to solve this problem, but especially with the popularity of fine-tuned transformers a simple way is to just encode both with the same vectorizer and use a delimiter or special token type to delineate the 2 as input to a BERT-like model that is then used to predict the class.

Updates to the infrastructure in this casting of the problem are minimal -- we need a reader for paired data (the most common formats for input are JSONL and TSV), and we need some sort of fine-tune classifier (or dual encoder).

This is a first attempt at that modification to mead without having to define a new task

### CODE REVIEW

File: baseline/tf/classify/training/distributed.py [hunk 1]
```diff
@@ -322,9 +332,9 @@ def fit_eager_distributed(model_params, ts, vs, es=None, **kwargs):
     if es is not None:
         print('Reloading best checkpoint')
         trainer.recover_last_checkpoint()
-        trainer.strategy = tf.distribute.OneDeviceStrategy('/device:GPU:0')
+        trainer.strategy = tf.distribute.OneDeviceStrategy('/device:CPU:0')
```
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/tf/classify/training/distributed.py:335]: Yeah, I have to figure out a better work-around.  Its fine to leave on the GPU in most cases, but not if we have a TPU running

File: baseline/tf/classify/training/distributed.py [hunk 2]
```diff
@@ -158,53 +159,62 @@ def _test(self, loader, steps=0, **kwargs):
...
 
         @tf.function
         def _distributed_test_step(inputs):
             per_replica_loss, per_replica_batchsz, per_replica_cm = strategy.experimental_run_v2(_replica_test_step, args=(inputs,))
             step_loss = strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_loss, axis=None)
             step_batchsz = strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_batchsz, axis=None)
-            step_cm = strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_cm, axis=None)
-            return step_loss, step_batchsz, step_cm
+            # step_cm
+            step_acc = strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_cm, axis=None)
```
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/tf/classify/training/distributed.py:188]: Right.  I was thinking of maybe having some options for which function we use.  The logic for multi-GPU works fine with `SparseTensor` but not on TPU.

File: baseline/pytorch/classify/model.py [hunk 3]
```diff
@@ -375,6 +375,72 @@ def forward(self, inputs):
...
+
+        toks = batch_dict[f"{key}"]
+        token_type_key = f"{key}_tt"
+        tt = batch_dict.get(token_type_key)
+        if numpy_to_tensor:
+            toks = torch.from_numpy(toks)
+            if tt is not None:
+                tt = torch.from_numpy(tt)
+        if self.gpu:
+            toks = toks.cuda()
```
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/pytorch/classify/model.py:407]: It probably should be, but right now the PyTorch code really needs an overhaul for multi-GPU from DataParallel to DistributedDataParallel.  I think its probably safe to punt this till that gets fixed

File: baseline/pytorch/classify/model.py [hunk 4]
```diff
@@ -375,6 +375,72 @@ def forward(self, inputs):
...
+
+    For the input, we will assume that the vectorizer will be producing a start token and an end token.
+    We will simply remove the start token from the second sentence and concatenate
+    [CLS] this is sentence one [SEP]
+
+    [CLS] this is sentence two [SEP]
+
+
+    """
+    def _convert_pair(self, key, batch_dict, numpy_to_tensor, example_dict):
```
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/pytorch/classify/model.py:397]: Potentially.  I just wanted to get it working for starters though, as you said.

File: baseline/pytorch/classify/model.py [hunk 5]
```diff
@@ -375,6 +375,72 @@ def forward(self, inputs):
         return self.output_layer(stacked)
 
 
+@register_model(task='classify', name='fine-tune-paired')
+class FineTunePairedClassifierModel(FineTuneModelClassifier):
+
+    """Fine-tuning model for pairs
+
+    This model encodes a pair as a single utterance using some encoding scheme defined in
+    convert_to_input which is fed directly into the fine-tuning model.
```
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/pytorch/classify/model.py:384]: Correct, Ill fix that

File: baseline/reader.py [hunk 6]
```diff
@@ -277,6 +277,18 @@ def load_examples(self, tsfile, src_vocabs, tgt_vocab, do_shuffle, src_sort_key)
...
+
+def _try_read_labels(**kwargs):
+    label_file = kwargs.get('label_file')
+    label_list = kwargs.get('label_list')
+    label2index = {}
+    if label_file:
+        pre_labels = Counter(chain(*_read_from_col(0, listify(label_file))))
+        label2index = {l: i for i, l in enumerate(pre_labels)}
+    elif label_list:
+        label2index = {k: i for i, k in enumerate(label_list)}
```
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/reader.py:289]: yes, my assumption was that we should continue to support the old label file way, but especially in cases where its a dataset from a 3rd party site with no label file and only a few labels, we can just inline them, and that would be assumed to be de-duped already of course then

File: baseline/reader.py [hunk 7]
```diff
@@ -650,6 +660,319 @@ def load(self, filename, vocabs, batchsz, **kwargs):
...
+        label_idx = len(self.label2index)
+        if isinstance(files, str):
+            if os.path.isdir(files):
+                base = files
+                files = filter(os.path.isfile, [os.path.join(base, x) for x in os.listdir(base)])
+            else:
+                files = [files]
+        vocab = {k: Counter() for k in self.vectorizers.keys()}
+
+        for file_name in files:
```
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/reader.py:792]: I dont remember when/if we encountered this case but a whole lot of datasets later, havent really seen it.  I think maybe we dont care?

File: baseline/reader.py [hunk 8]
```diff
@@ -650,6 +660,319 @@ def load(self, filename, vocabs, batchsz, **kwargs):
...
+    def index_pair_label(self, line, clean_fn, header):
+        cols = line.strip().split('\t')
+        header = [h.strip() for h in header.split('\t')]
+        col_indices = [header.index(c) for c in self.col_keys]
+        # QQP files have some invalid samples
+        #if max(col_indices) > len(cols):
+        #    raise Exception(f"Error: {cols}")
+        index_pair_label_list = [cols[c] for c in col_indices]
+        index, text1, text2, label = index_pair_label_list
+        text1 = ' '.join(list(filter(lambda s: len(s) != 0, [clean_fn(w) for w in text1.split()])))
```
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/reader.py:955]: probably yeah

File: baseline/tf/classify/model.py [hunk 9]
```diff
@@ -635,6 +635,73 @@ def call(self, inputs):
...
+            #if not eager:
+            #    raise Exception("We arent currently supporting non-eager mode with token_types")
+            #else:
+            example_dict[target_key] = (toks, tt)
+        else:
+            example_dict[target_key] = toks
+
+    def call(self, inputs):
+        inputs_for_model = {}
+        for key in self.embeddings.keys():
```
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/tf/classify/model.py:674]: opened a separate ticket

File: baseline/reader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/reader.py]: updated var name to `l`

File: baseline/pytorch/classify/model.py [hunk 10]
```diff
@@ -375,6 +375,72 @@ def forward(self, inputs):
...
+    def make_input(self, batch_dict, perm=False, numpy_to_tensor=False):
+
+        """Transform a `batch_dict` into something usable in this model
+
+        :param batch_dict: (``dict``) A dictionary containing all inputs to the embeddings for this model
+        :return:
+        """
+        example_dict = dict({})
+        perm_idx = None
+        # Allow us to track a length, which is needed for BLSTMs
```
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/pytorch/classify/model.py:425]: removed

File: baseline/reader.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dpressel) [reply baseline/reader.py:955]: slight difference in implementation, which also requires a function pointer.  Im not sure if it makes sense to keep `splits` but seems like a separate issue that should be considered independently
